### DISCUSSION TITLE
Implementing an additional, country-independent language layer

### DISCUSSION THREAD

**Discussion by marcelbrode** [shopware/platform#5286] [RFC / Proposal]
## Current state

Currently, Shopware’s language system uses a specific language locale (e.g. `de-DE` for German in Germany) for its translation, while falling back to another specific locale (e.g. `en-GB` for English in Great Britain).

This works well, but it is quite maintenance-heavy. If you use multiple versions or dialects of a language, such as British and American English, it will result in a lot of duplicated snippets.

### Some facts and figures

- Shopware has approximately 12,000 snippets.
- By default, the US translation is an automated copy of the GB translation, which is later adjusted by our community via Crowdin.
- When running them through a comparison tool, we find about 70 differences. This means that roughly 99.5% of our American snippets are identical to the British ones.

## Solution

Another layer containing a language without a country context (e.g., `en` for both `en-GB` and `en-US`) would provide an additional fallback layer. This means dialects will be treated as patch files. In Shopware, `en-GB` would be renamed to `en`, as you have to select a base dialect, retaining its 12,000 snippets, while `en-US` would shrink to around 70 snippets. This will also eliminate the need for tools to initially copy-paste or pre-translate from "English to English," which, of course, incurs costs.

Many translation tools, such as vue-i18n or Symfony, already offer this functionality, but due to our extensible architecture, it was disabled during the development process.

Fortunately, this can be easily re-implemented without breaking anything, as demonstrated in my proof of concept.

### Proof of concept
[Add additional language layer to core](https://github.com/shopware/shopware/pull/5138/files#diff-4dcf5c3effda23eb35225683ee6d0c613dfe27f4058983a5aebf3392a3300d2f)

## Benefits & consequences

- Every additional dialect’s snippet file (e.g., `en-US` for `en`) will shrink from around 12,000 entries to likely fewer than 100.
- Translations provided via the community (Crowdin) will show a lower coverage percentage, but will more accurately reflect the actual differences between `en-US` and `en`.
- In Crowdin and in our product, the base language is set declaratively, so we need to inform the community that British English is our base language, and we simply refer to it as `en`.
- All snippet files in Shopware’s core will be renamed to the base language
    - `en-GB.json` -> `en.json`, `messages.en-GB.base.json` -> `messages.en.base.json` and so on.
    - Similarly: `de-DE` -> `de`
- The language pack will update its locales to their base versions as well (e.g., `es-ES` -> `es`).
- The system’s default language or active language handling will remain unchanged.
- Plugins and apps do not require immediate adjustments, as the previous structure will continue to function as before.

### How will compatibility with existing plugins and extensions be ensured?
As hinted in the consequences, we can just rename the language files, and that's it. In the meantime, everything works as before; it just doesn't have the `en` fallback. That's one of the really nice things about this POC.

### Will there be a migration script that automatically converts en-GB to en?
Currently not, but I like the idea. It will definitely be added later on! For the core, it's not necessary, since everything is renamed already.

### How will the new system affect the community’s work on Crowdin?
It won't. Some might be surprised that the coverage will go down, but that's the only noticeable change in Crowdin.

### How exactly will the new fallback system be implemented technically?
To be blunt, there will be one extra file to be loaded and merged into the active "snippet catalogue" before the specific patch file (the current standard) is added. Symfony does this out of the box, but this has to be adjusted because it was kind of disabled when we implemented the database snippet overrides back then. Feel free to have a look at my [POC](https://github.com/shopware/shopware/pull/5138/files#diff-4dcf5c3effda23eb35225683ee6d0c613dfe27f4058983a5aebf3392a3300d2f) as well.

### Why is it important to shrink snippet file sizes?
Mostly for maintainance reasons. When an English snippet changes, there has to be manual validation of every change, which in 99% of cases is exactly the same as British English. From a performance perspective, it seems to not have any real impact so far. In the storefront, snippet loading (old and new) only takes 1-2ms, and in the administration, 50-100ms in both cases. But I'll probably investigate this more later on.

### With our current US focus, shouldn't the default be en-US?
I see the point, but that would be another topic. Changing the system's default language is a whole different matter and should be discussed at another point, since, in my opinion, this is out of scope.

### What exactly is the deprecation strategy?
That's the neat part: there's none. This is not a new system. It's an extension of the old system, "just" pushing another layer in-between. Plugins or apps that do not notice this change will work the same, while with the active `en-GB` language, the core will only 100% fallback to en.

### Wait a minute, I thought language fallback for dialects already exists?
You're probably talking about [content language inheritance](https://docs.shopware.com/en/shopware-6-en/settings/languages#add-a-new-language). That is something different. That feature covers translating content like product names. I am talking about snippet translations, like the whole admin or storefront UI. Everything **but** content, to be precise.

### Current 

```mermaid
flowchart LR
    subgraph "Specific language"
        deDE("<strong>de-DE</strong><br/><em>German in Germany</em>")
        deAT("<strong>de-AT</strong><br/><em>German in Austria</em>")
        esES("<strong>es-ES</strong><br/><em>Spanish in Spain</em>")
        esAR("<strong>es-AR</strong><br/><em>Spanish in Argentina</em>")
        enUS("<strong>en-US</strong><br/><em>English in the USA</em>")
    end
    subgraph "System default language"
        enGB("<strong>en-GB</strong><br/><em>English in Great Britain</em>")
    end
        
    deDE-->enGB
    deAT-->enGB

    esES-->enGB
    esAR-->enGB
    
    enUS-->enGB
```

### Target

```mermaid
flowchart LR
    subgraph "Specific language (German)"
        deDE("<strong>de-DE</strong><br/><em>German in Germany</em>")
        deAT("<strong>de-AT</strong><br/><em>German in Austria</em>")
        de("<strong>de</strong><br/><em>German</em>")
    end
    subgraph "Specific language (Spanish)"
        esES("<strong>es-ES</strong><br/><em>Spanish in Spain</em>")
        esAR("<strong>es-AR</strong><br/><em>Spanish in Argentina</em>")
        es("<strong>es</strong><br/><em>Spanish</em>")
    end
    subgraph "Specific language (English)"
        enGB("<strong>en-GB</strong><br/><em>English in Great Britain</em><br\> 
 <em>(Will have no files)</em>")
        enUS("<strong>en-US</strong><br/><em>English in the USA</em>")
    end
    subgraph "System default language"
        en("<strong>en</strong><br/><em>English</em>")
    end

    deDE-->de
    deAT-->de
    de-->en

    esES-->es
    esAR-->es
    es-->en

    enGB-->en
    enUS-->en

    style de color:#6f6,stroke:#6f6
    style es color:#6f6,stroke:#6f6
    style en color:#6f6,stroke:#6f6
    
    style enGB stroke-dasharray: 5
```

**Comment by keulinho** [comment]
Just had a funny discussion with @tobiasberge around this, and actually this is still super recent and valid and solves a lot of the problems we have regarding translation handling. We really like the idea 👍 

We only have two questions in addition to what is stated here:
* We are not sure if it really is not a break for extensions:
Before an extension needed to ship only en-GB snippets, this means when another language is used (e.g. de-DE) that the extension does not provide it will fallback to en-GB, which is not nice, but should be the most usable solution in that case.
However when that plugin in the new system exists, it would look into de-DE, which is not defined for the snippet, then it would look for de, also not defined, and then for en which is also not defined for the given snippet. Based on the information here I would assume in that case it will only display the snippet key as we don't find any value in the translation chain, I would consider that a break.

So we would need (maybe temporarily) to have a fallback from en to en-GB or something

* How would this look in the admin where you can override snippets, do we allow that for the base languages as well or only for the specific localized languages?

**Reply by marcelbrode** [reply]
Thanks for you input! Lemme get right into it:
- Totally valid point. That's at least how I described it. I'm not 100% sure, but I THINK 'en-GB' will also be in-between. If not, this should at least be a case we should think of (especially for english). Our team wants to evaluate next steps soon anyway, so I'll put this. Maybe some kind of file migration could do the trick 🤔 
- You're right. Snippet Module overrides should have some kind of inheritance logic as well. I'll create a ticket for that as well.

I'll update you if I know more 😄

**Reply by marcelbrode** [reply]
@keulinho (CC: @MartinKrzykawski)
I had a look at it and luckily, the inheritance is not really a problem!
The thing is how the admin snippets are working:

It loads the translation of the current language and the fallback language anyway. Since my POC affects the snippet finder, it will still request "en-GB" and will now get the merge of "en" and "en-GB" instead of just "en-GB". So the Admin inheritance is now:


```mermaid
flowchart LR
    subgraph "Specific language (German)"
        deDE("<strong>de-DE</strong><br/><em>German in Germany</em>")
        de("<strong>de</strong><br/><em>German</em>")
    end
    subgraph "Fallback (English)"
        enGB("<strong>en-GB</strong><br/><em>English in Great Britain</em>")
        enUS("<strong>en-US</strong><br/><em>English in the USA</em>")
        en("<strong>en</strong><br/><em>English</em>")
    end

    deDE-->de
    de-->enGB
    enGB-->en
    enUS-->en
    
    style enGB stroke-dasharray: 5
```

To be identical, we will talk in the refinement today, if we wanna explicitly add en-GB to the Storefront fallback as well. Since the Storefront didn't have any fallback before, this isn't a break anyway.

**Reply by keulinho** [reply]
Thanks for the detailed answer, that already explains a lot 👍 

I guess we would need the explicit en-gb fallback (or implicit migration to en base if no base is given in extensions) -> the case I have in mind right now:
Plugin only ships `de-DE`and `en-GB` snippets, user wants `en-US` -> before that would use the en-GB snippets which is expected -> by default now en-US would only fallback to en which is not defined

So for me it makes most sense in extension to treat en-GB as base en if not explicit en base file is present -> I guess this is similar to what you have in mind when you say `Maybe some kind of file migration could do the trick` right?

**Reply by marcelbrode** [reply]
So, refinement is done. We will also add the en-GB layer to the storefront, but with an instant deprecation for 6.8, so that Plugin Devs have a year to adjust their translations.

**Reply by JoshuaBehrens** [reply]
@marcelbrode how has this snippet migration to be done by devs? Can we still ship en-GB and en or do we have to make sure to have a separate build of assets, that explicitly does not include a certain translation? Do we need to provide migrations for overrides already happening in the database?

**Reply by marcelbrode** [reply]
> @marcelbrode how has this snippet migration to be done by devs? Can we still ship en-GB and en or do we have to make sure to have a separate build of assets, that explicitly does not include a certain translation?

As soon as the POC has been implemented, the devs should rename their files from e.g. "en-GB" to "en", since the base language is now the important one. If I get your question correctly, there's absolutely no need for seperate builds of assets. If the "activated" language is english, everything will still work out of the box, since en-GB will always fallback to en. It doesn't matter in this case! The only matter is the symfony translator for fallbacks from another language. Following scenario:
- Translations in `en-GB` instead of `en`
- Something has NOT been transalted in german
- In 6.8, Fallback will skip en-GB from german and fall to english. In that case you won't have translations

> Do we need to provide migrations for overrides already happening in the database?

Database translations are not affected since they were always mapped to a specific language and are in the highest level of translation. This won't change.

**Comment by marcelbrode** [comment]
We can close this discussion, since we'll do this 😄 
https://github.com/shopware/shopware/pull/5138
